flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -128,6 +130,20 @@
 	/** Resource profiles of slots that can be acquired. */
 	private final HashSet<ResourceProfile> availableSlotResourceProfiles = new HashSet<>();
 
+	/**
+	 * All allocated slots's allocation resource profile.
+	 * If the task manager has no allocated slots(all slots are free), it does not exist in this map.
+	 * ResourceID : The id of each task manager
+	 * Tuple2 : The allocated slots and allocation resources of the task manager
+	 *    f0 : The allocated slots map, keyed by slotId and value is allocation resource profile of the slot
+	 *    f1 : The remaining resource of the task manager identified by resourceId
+	 */
+	private final Map<ResourceID, Tuple2<Map<SlotID, ResourceProfile>, ResourceProfile>>
```

Comment:
I think it may be better to have a distinct class to manage the total resource, the free resource, the resource of the allocated slot. Then we could merge these two map with a simpler map.